runtime-sdk/modules/evm/src/signed_query.rs

```diff
@@ -0,0 +1,248 @@
+        "(",
+        "address from",
+        ",address to",
+        ",uint256 value",
+        ",uint256 gasPrice",
+        ",uint64 gasLimit",
+        ",bytes data",
+        ",Leash leash",
+        ")",
+        leash_type_str!()
+    );
+    hash_encoded(&[
+        encode_bytes(CALL_TYPE_STR),
+        Token::Address(query.caller.0.into()),
+        Token::Address(query.address.0.into()),
+        Token::Uint(ethabi::ethereum_types::U256(query.value.0)),
+        Token::Uint(ethabi::ethereum_types::U256(query.gas_price.0)),
+        Token::Uint(query.gas_limit.into()),
+        encode_bytes(&query.data),
+        Token::Uint(hash_leash(query.leash.as_ref().unwrap() /* checked above */).into()),
+    ])
+}
+
+fn hash_leash(leash: &Leash) -> [u8; 32] {
+    hash_encoded(&[
+        encode_bytes(leash_type_str!()),
+        Token::Uint(leash.nonce.into()),
+        Token::Uint(leash.block_number.into()),
+        Token::Uint(leash.block_hash.0.into()),
+        Token::Uint(leash.block_range.into()),
+    ])
+}
+
+fn hash_domain<Cfg: Config>() -> &'static [u8; 32] {
+    static DOMAIN_SEPARATOR: OnceCell<[u8; 32]> = OnceCell::new(); // Not `Lazy` because of generic.
+    DOMAIN_SEPARATOR.get_or_init(|| {
+        const DOMAIN_TYPE_STR: &str = "EIP712Domain(string name,string version,uint256 chainId)";
+        hash_encoded(&[
+            encode_bytes(DOMAIN_TYPE_STR),
+            encode_bytes("Sapphire ParaTime"),
+            encode_bytes("1.0.0"),
+            Token::Uint(Cfg::CHAIN_ID.into()),
+        ])
+    })
+}
+
+fn encode_bytes(s: impl AsRef<[u8]>) -> Token {
+    Token::FixedBytes(Keccak256::digest(s.as_ref()).to_vec())
+}
+
+fn hash_encoded(tokens: &[Token]) -> [u8; 32] {
+    Keccak256::digest(&ethabi::encode(tokens)).into()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use oasis_runtime_sdk::testing::mock;
+
+    use crate::{test::ConfidentialEVMConfig as Cfg, types::SignedQueryEnvelope};
+
+    /// This was generated using the `@oasislabs/sapphire-paratime` JS lib.
+    const SIGNED_QUERY: &str =
+"a2657175657279a764646174614401020304656c65617368a4656e6f6e63651903e76a626c6f636b5f686173685820c92b675c7013e33aa88feaae520eb0ede155e7cacb3c4587e0923cba9953f8bb6b626c6f636b5f72616e6765036c626c6f636b5f6e756d626572182a6576616c75655820000000000000000000000000000000000000000000000000000000000000002a6663616c6c65725411e244400cf165ade687077984f09c3a037b868f676164647265737354b5ed90452aac09f294a0be877cbf2dc4d55e096f696761735f6c696d69740a696761735f70726963655820000000000000000000000000000000000000000000000000000000000000007b697369676e617475726558412496b1b40cfa931d6e62f76148bdb37e0a775213d757d715d379e60dfadfd18602296c8480d117fce51a52e2e420dbc6017d1e42accb0e5393a78050fcca73d11c";
+
+    fn setup_context<C: Context>(ctx: &mut C, query: &SimulateCallQuery, omit_block: bool) {
+        let leash = query.leash.as_ref().unwrap();
+        let mut state = ctx.runtime_state();
+
+        // Set the nonce to what the leash requires.
```

Comment:
Add a test with bad nonce.

Reply:
Thanks. I submitted this patch at 3 am. I could barely keep my eyes open, and I had the feeling I was missing something :)